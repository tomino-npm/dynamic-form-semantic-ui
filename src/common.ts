import { FormElement, DataSet } from '@tomino/dynamic-form';
import { initCss } from '@tomino/toolbelt';

export type FormControlProps = {
  formControl: FormElement;
  owner: DataSet;
  handlers?: { [index: string]: (...params: any[]) => any };
};

export const { css, renderCss } = initCss('d-form');
