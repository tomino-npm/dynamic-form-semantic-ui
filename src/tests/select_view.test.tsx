import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { create } from './form_query_data';
import { JSONSchema, FormDefinition, FormModel, config } from '@tomino/dynamic-form';
import { TestComponent } from './common';

describe('Form', function() {
  const schema: JSONSchema = {
    type: 'object',
    properties: {
      country: {
        type: 'string'
      },
      city: {
        type: 'string'
      },
      countries: {
        $enum: [{ text: 'Australia', value: 'AU' }, { text: 'Slovakia', value: 'SK' }]
      },
      cities: {
        $enum: [
          { text: 'Sydney', value: 'SYD', country: 'AU' },
          { text: 'Melbourne', value: 'MEL', country: 'AU' },
          { text: 'Kosice', value: 'KE', country: 'SK' },
          { text: 'Bratislava', value: 'BA', country: 'SK' }
        ]
      }
    }
  };

  const controlData = { country: 'SK', city: 'KE' };

  const formDefinition: FormDefinition = create.form({
    elements: [
      create.formElement({
        row: 0,
        column: 0,
        width: 8,
        list: 'countries',
        control: 'Select',
        label: 'Country',
        source: 'country',
        inline: true
      }),
      create.formElement({
        row: 0,
        column: 8,
        width: 8,
        control: 'Select',
        list: 'cities',
        filterSource: 'country',
        filterColumn: 'country',
        controlProps: {
          search: true
        },
        source: 'city',
        label: 'City'
      })
    ]
  });

  describe('Select', function() {
    const setDirty = jest.fn();

    beforeAll(() => {
      config.setDirty = () => setDirty();
    });

    function componentWithData() {
      const form = new FormModel(formDefinition, schema, controlData);

      // just another notation
      return <TestComponent form={form} />;
    }

    it('renders correctly', () => {
      const component = renderer.create(componentWithData());
      expect(component).toMatchSnapshot();
    });

    it('changes value and all related formulas', () => {
      const component = renderer.create(componentWithData());
      const root = component.root;
      const country = root.findByProps({ name: 'country' });
      country.props.onChange(null, { value: 'AU' });

      const city = root.findByProps({ name: 'city' });
      city.props.onChange(null, { value: 'SYD' });

      expect(component).toMatchSnapshot();
      expect(setDirty).toBeCalled();
    });

    return { componentWithData };
  });
});
