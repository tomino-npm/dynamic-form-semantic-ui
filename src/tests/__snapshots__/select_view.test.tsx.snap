// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Select changes value and all related formulas 1`] = `
<div
  className="ui segment ui form"
>
  <style>
    .d-form-bgcibidbjc { 
  color: pink;
  font-size: 100px !important;
 }
.d-form-Aebfdiadfh { 
  padding: 9px 0px;
 }
.d-form-Adiagibjfa { 
  margin-right: 12px;
  margin-bottom: 6px;
 }
.d-form-bidfabidci { 
  margin: 6px 0px 12px 0px;
 }
.d-form-Abdgbhfjheg { 
  margin-bottom: 0px !important;
 }

  .fieldSet-bejaheefjb {
    border: solid 1px #ddd;
    border-radius: 5px;
    background-color: #fafafa;
  }


  .fieldSet-bejaheefjb legend {
    padding: 0px 6px;
    font-weight: bold;
  }
  </style>
  <div
    className="ui form"
  >
    <div
      className="fields"
    >
      <div
        className="inline field"
      >
        <label
          htmlFor="country"
        >
          Country
        </label>
        <div
          aria-expanded={false}
          className="ui selection dropdown"
          name="country"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseDown={[Function]}
          role="listbox"
          tabIndex={0}
        >
          <div
            aria-live="polite"
            className="text"
            role="alert"
          >
            Australia
          </div>
          <i
            aria-hidden="true"
            className="dropdown icon"
            onClick={[Function]}
          />
          <div
            className="menu transition"
          >
            <div
              aria-checked={true}
              aria-selected={true}
              className="active selected item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                Australia
              </span>
            </div>
            <div
              aria-checked={false}
              aria-selected={false}
              className="item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                Slovakia
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        className="eight wide field"
      >
        <label
          htmlFor="city"
        >
          City
        </label>
        <div
          aria-expanded={false}
          className="ui search selection dropdown"
          name="city"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseDown={[Function]}
          role="combobox"
        >
          <input
            aria-autocomplete="list"
            autoComplete="off"
            className="search"
            onChange={[Function]}
            style={
              Object {
                "width": undefined,
              }
            }
            tabIndex={0}
            type="text"
            value=""
          />
          <div
            aria-live="polite"
            className="text"
            role="alert"
          >
            Sydney
          </div>
          <i
            aria-hidden="true"
            className="dropdown icon"
            onClick={[Function]}
          />
          <div
            className="menu transition"
            role="listbox"
          >
            <div
              aria-checked={true}
              aria-selected={true}
              className="active selected item"
              country="AU"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                Sydney
              </span>
            </div>
            <div
              aria-checked={false}
              aria-selected={false}
              className="item"
              country="AU"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                Melbourne
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Form Select renders correctly 1`] = `
<div
  className="ui segment ui form"
>
  <style>
    .d-form-bgcibidbjc { 
  color: pink;
  font-size: 100px !important;
 }
.d-form-Aebfdiadfh { 
  padding: 9px 0px;
 }
.d-form-Adiagibjfa { 
  margin-right: 12px;
  margin-bottom: 6px;
 }
.d-form-bidfabidci { 
  margin: 6px 0px 12px 0px;
 }
.d-form-Abdgbhfjheg { 
  margin-bottom: 0px !important;
 }

  .fieldSet-bejaheefjb {
    border: solid 1px #ddd;
    border-radius: 5px;
    background-color: #fafafa;
  }


  .fieldSet-bejaheefjb legend {
    padding: 0px 6px;
    font-weight: bold;
  }
  </style>
  <div
    className="ui form"
  >
    <div
      className="fields"
    >
      <div
        className="inline field"
      >
        <label
          htmlFor="country"
        >
          Country
        </label>
        <div
          aria-expanded={false}
          className="ui selection dropdown"
          name="country"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseDown={[Function]}
          role="listbox"
          tabIndex={0}
        >
          <div
            aria-live="polite"
            className="text"
            role="alert"
          >
            Slovakia
          </div>
          <i
            aria-hidden="true"
            className="dropdown icon"
            onClick={[Function]}
          />
          <div
            className="menu transition"
          >
            <div
              aria-checked={false}
              aria-selected={false}
              className="item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                Australia
              </span>
            </div>
            <div
              aria-checked={true}
              aria-selected={true}
              className="active selected item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                Slovakia
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        className="eight wide field"
      >
        <label
          htmlFor="city"
        >
          City
        </label>
        <div
          aria-expanded={false}
          className="ui search selection dropdown"
          name="city"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseDown={[Function]}
          role="combobox"
        >
          <input
            aria-autocomplete="list"
            autoComplete="off"
            className="search"
            onChange={[Function]}
            style={
              Object {
                "width": undefined,
              }
            }
            tabIndex={0}
            type="text"
            value=""
          />
          <div
            aria-live="polite"
            className="text"
            role="alert"
          >
            Kosice
          </div>
          <i
            aria-hidden="true"
            className="dropdown icon"
            onClick={[Function]}
          />
          <div
            className="menu transition"
            role="listbox"
          >
            <div
              aria-checked={true}
              aria-selected={true}
              className="active selected item"
              country="SK"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                Kosice
              </span>
            </div>
            <div
              aria-checked={false}
              aria-selected={false}
              className="item"
              country="SK"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                Bratislava
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
